### ISSUE TITLE
[Feature Request] language injection 

### ISSUE THREAD

**Issue by aMOPel** [alaviss/tree-sitter-nim#24]
Your progress is very impressive. Thanks for your hard work!

I saw that you proceeded to make the first queries for highlighting. And since you're still optimizing a bunch of stuff, I thought I'd make you aware of a feature that would be great to have, so you can consider it in your optimization considerations.

[Injection queries](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#injections) let you inject other TS parsers at certain points in your language.

This would be very useful in the following locations:
- strings:
  - `sql""` https://nim-lang.org/docs/db_common.html
  - `re""` / `rex""` https://nim-lang.org/docs/re.html
  - `peg""` https://nim-lang.org/docs/pegs.html
- asm statement https://nim-lang.org/docs/manual.html#statements-and-expressions-assembler-statement
- comments:
  - reStructured text in doc comments
  - https://github.com/stsewd/tree-sitter-comment in normal comments
- pragmas:
  - emit pragma https://nim-lang.org/docs/manual.html#implementation-specific-pragmas-emit-pragma
  - (importc/Cpp/ObjC/JS pragma ) those have their own special syntax 

Those are all foreign languages, that I am aware of, that are used inside of Nim.

Additionally to injecting those it would be great to have support for format strings:
- `fmt""` and `&""` https://nim-lang.org/docs/strformat.html

While the ASM statement and the comments may be trivial to query for, querying for the specific generalized raw string literals and the emit pragma most certainly requires to use regex in the query. 

This would normally not be a problem, however nvim seems to have [performance issues with injection queries](https://github.com/neovim/neovim/issues/22426), so keeping them as lean as possible would be preferable. My proposal would be to make the querying cheaper, by letting the parser do the work and perse them as specific ast nodes.

**Comment by alaviss** [comment]
> The emit pragma needs a preceding comment to specify the emitted language eg.

I'd say use C unless specified otherwise.

> Do you have suggestions or other ideas for injections?

I think you got most of it.

**Comment by aMOPel** [comment]
> Gonna try.

So, I don't think it's a good idea. The case without the comment (CPP) is more generic, CPP also gets injected on top of the other language.
When 2 languages get injected, it just puts both in the AST, which means highlights get wonky, as the CPP parser gives many errors when parsing a different language.

**Comment by aMOPel** [comment]
> You think string interpolation is feasible?

A while back I worked a little on string interpolation in one of my parser rewrite attempts. It worked in my grammar.

```javascript
{
    // &"" is a str_lit but fmt"" is a generalized_str_lit and thus raw
    interpolated_string_lit: $ => choice(
      prec(PREC.interpolated_string_lit, seq(
        alias('&', $.interpolated_string_prefix),
        field('open', '"'),
        repeat(choice(
          token.immediate(/[^\n\r"]/),
          $.str_esc_seq,
          $.string_interpolation,
        )),
        field('close', token.immediate('"')),
      )),
      seq(
        alias('fmt', $.interpolated_string_prefix),
        field('open', token.immediate('"')),
        optional(seq(
          token.immediate(/[^\r\n"]/), // to disambiguate from GENERALIZED_TRIPLESTR_LIT
          repeat(choice(
            token.immediate(/[^\r\n"]|""/),
            $.string_interpolation,
          )),
        )),
        field('close', token.immediate('"')),
      ),
    ),

    string_interpolation: $ => seq(
      field('open', token.immediate('{')),
      $.expr,
      repeat(token.immediate(' ')),
      field('close', token.immediate('}')),
    ),
}
```

The annoying thing was the differentiation between the `&` and `fmt` prefix, since one is a string prefixed by an operator and the other is a generalized string.

For the multiline string, it doesn't matter so much, since that is always 'raw".

**Comment by alaviss** [comment]
> You think string interpolation is feasible?

I think for sanity reasons this might work out better as an injection query of sorts.

It also means you'd have to write an external grammar for injections so maybe not that great of an idea.

**Comment by alaviss** [comment]
> The case without the comment (CPP) is more generic, CPP also gets injected on top of the other language.

One of the few times you want conditionals, I guess.

Though I think it might be worth considering to just inject C/C++ syntax always because most code doesn't have the comment + language thing.

**Comment by alaviss** [comment]
> One of the few times you want conditionals, I guess.

I think you should investigate whether it is possible to have queries for language variants.

Since Nim in JS mode also have some quirks like `asm` emitting JS which is (ab)used heavily by `jsffi` IIRC.

**Comment by aMOPel** [comment]
> language variants

Hm. So something like a `.c.nim` file, which becomes a `nim.c` filetype in vim?

**Comment by aMOPel** [comment]
> It also means you'd have to write an external grammar for injections so maybe not that great of an idea.

Since there can be any simple expression in the `{}` it would just be a subset of your grammar. So duplication and doubled maintenance.

You don't think it's easily doable in this your grammar?

**Comment by alaviss** [comment]
> Hm. So something like a `.c.nim` file, which becomes a `nim.c` filetype in vim?

That could be a thing. Might have to check with neovim folks though.

> Since there can be any simple expression in the `{}` it would just be a subset of your grammar.

I'm proposing a double injection:

- With querying it's trivial to query for `fmt("string")` or `fmt"string"`, same as with `&`. Use that to inject `nim-fmt`.
- In `nim-fmt`, make `{}` turns into a node that you then inject `nim` back into it.

I'll admit that I haven't tested injecting this grammar into anything, so the external scanner might need some love, but hopefully not.

> You don't think it's easily doable in your grammar?

I think this is simpler than trying to tackle it within this grammar and handle the various conflicts with the various forms of call expression, which is already being held together by a very carefully crafted set of conflict resolutions.

**Comment by aMOPel** [comment]
I see. That could work. However then I'm gonna end up putting 2 grammars and 3 language variants into the nvim treesitter repo just for Nim :D gonna see if that's alright for the nvim treesitter maintainers.

**Comment by aMOPel** [comment]
Just had another idea. A compromise.

We turn matching `{}` into a named node always for all strings in this grammar. Then I query for fmt/& and inject this grammar inside again.

No extra grammar needed and the modification to this grammar shouldn't have conflicts with other stuff right?

**Comment by aMOPel** [comment]
Alright I'm gonna try the extra grammar idea

**Comment by aMOPel** [comment]
So I got this now:

```scheme
module.exports = grammar({
  name: 'nim_format_string',

  extras: $ => [
    /[\n\r\s]+/,
  ],

  rules: {
    source_file: $ => optional($.string_content),
    string_content: $ => repeat1(choice($.matching_curlies, $._everything_else)),
    _everything_else: $ => /[^\n\r\s]/,
    matching_curlies: $ => seq(
      '{',
      alias(repeat($._everything_else), $.everything_else_inside),
      '}',
    ),
  }
});
```

It works alright. The main issue is with the greedy matching. If it encounters a `}` inside a matching `{}` eg. `{ {'a'..'z'} }`, it will close the curlies prematurely. You got any smart ideas how to solve this in the grammar.js? Or is that only doable by the scanner.cc?

**Comment by aMOPel** [comment]
https://github.com/aMOPel/tree-sitter-nim-format-string

**Comment by alaviss** [comment]
> Or is that only doable by the scanner.cc?

Yea, you'd need some state to track inner curlies if you want one flat node.

However, you can try a recursive grammar to match inner curlies. It'd generate some inner nodes so hopefully that won't impact injection.

**Comment by aMOPel** [comment]
> However, you can try a recursive grammar to match inner curlies. It'd generate some inner nodes so hopefully that won't impact injection.

It could also just be the closing curly though, eg `{ "}" }`. silly example, but you get my meaning.

**Comment by alaviss** [comment]
Try looking at the parser itself to create rules then:

https://github.com/nim-lang/Nim/blob/38b58239e882eaa905bafa49237f0b9ca9d43569/lib/pure/strformat.nim#L583

It's pretty short and concise.

**Comment by aMOPel** [comment]
Translate that to cpp you mean? I'm honestly pretty bad at it.

**Comment by alaviss** [comment]
> Translate that to cpp you mean? I'm honestly pretty bad at it.

No, what I meant is to conjure out grammar nodes for tree-sitter.

**Comment by alaviss** [comment]
I guess something like this?

```js
interpreted: $ => seq(
  '{',
  repeat(
    choice(
      /[^\\'"({]/,
      $._escaped,
      $._doublequotes,
      $._singlequotes,
      $._parens,
      $.interpreted
    )
  ),
  '}'
),
```

**Comment by aMOPel** [comment]
I read some into the docs https://nim-lang.org/docs/strformat.html.

There is actually way more to it than I thought.

```nim
assert fmt"{123.456:>9.3f}" == "  123.456"             # format parameters after `:`  not normal nim expressions
assert fmt"{123.456=}" == "123.456=123.456"            # `=` has special meaning
assert fmt"{123.456=:>9.3f}" == "123.456=  123.456"    # `=` and : together
```

There is a whole mini syntax for those format parameters: https://nim-lang.org/docs/strformat.html#standard-format-specifiers-for-strings-integers-and-floats

Also
> Parts of the string that are enclosed in the curly braces are interpreted as Nim code. To escape a { or }, double it.
Within a curly expression, however, {, }, must be escaped with a backslash.

So `{{` outside curlies are just `{` in the string, BUT `{}` need to be escaped `\{\}` within curlies to be interpreted as normal `{}`. That means the "normal" Nim expressions inside curlies are not even normal. Eg
`fmt"{ \{'a'..'z'\} }"` is the way to write a set inside curlies. So injecting the normal Nim parser won't even work, since it doesn't know what to do with those `\` and also I can't throw away the `\` before passing it to the nim parser either.

Basically what I'm saying is, that a full support of the spec seems impossible to me rn.

**Comment by alaviss** [comment]
> Basically what I'm saying is, that a full support of the spec seems impossible to me rn.

I think let's just focus on what we can do.

Maybe you can detect the ones with escapes and disable injection.

**Comment by aMOPel** [comment]
Pushed my work on the special syntax.

https://github.com/aMOPel/tree-sitter-nim-format-string/commit/d678bca42c8b424a57938cfca1d9ad787ff68a2f#diff-919ac210accac9ecc55a76d10a7590e3d85ca3f0e165b52d30f08faee486d0cb

Your wisdom on grammar would be appreciated :)
